### PR TITLE
Create wrapper structs and interfaces for properties data structures and payloads

### LINKED ISSUE #1239: Revisit return values of unicode property functions
(there's probably an issue for this already, but I'm filing it to make sure)

Currently we have

```rust
pub fn get_ascii_hex_digit<'data, D>(provider: &D) -> UnisetResult<'data>

pub fn get_script<'data, D>(provider: &D) -> CodePointMapResult<'data, Script>
```

which results in call sites that look like

```rust
// Sets
let provider = icu_testdata::get_provider();
let payload =
    sets::get_ascii_hex_digit(&provider)
        .expect("The data should be valid");
let data_struct = payload.get();
let ascii_hex_digit = &data_struct.inv_list;
assert!(ascii_hex_digit.contains('3'));

// Maps
let provider = icu_testdata::get_provider();
let payload =
    maps::get_script(&provider)
        .expect("The data should be valid");
let data_struct = payload.get();
let script = &data_struct.codepoint_trie;
assert_eq!(script.get('木' as u32), Script::Han);  // U+6728
```

We should think about this further.

CC @echeran @markusicu

### PR DESCRIPTION
**Pull request by echeran**
Fixes #1239

### CODE REVIEW

File: CODEOWNERS [hunk 1]
```diff
@@ -14,7 +14,7 @@ components/icu4x/                 @unicode-org/icu4x-owners
 components/locale_canonicalizer/  @dminor @zbraniecki
 components/locid/                 @zbraniecki @nciric
 components/plurals/               @zbraniecki @sffc
-components/uniset/                @echeran @iainireland
+components/codepointset/          echeran @iainireland
```
Comment (sffc) [inline CODEOWNERS:17]: Please restore yourself

File: .github/workflows/build-test.yml [hunk 2]
```diff
@@ -468,7 +468,7 @@ jobs:
           - components/locid
           - components/plurals
           - components/datetime
-          - utils/uniset
+          - utils/codepointset
```
Comment (sffc) [inline .github/workflows/build-test.yml:471]: Reminder to update the gh_pages branch

File: docs/process/prd.md [hunk 3]
```diff
@@ -57,7 +57,7 @@ ICU4X 0.1’s selection of components aimed to:
 
 -   Validate low-level models around [data management](https://github.com/unicode-org/icu4x/blob/master/docs/design/data_pipeline.md)
 -   Establish project culture using simple building blocks such as [Locale](https://docs.rs/icu_locid/0.1.0/icu_locid/) and [Plural Rules](https://docs.rs/icu_plurals/0.1.0/icu_plurals/)
--   Introduce a low-level foundation for string operations with [UnicodeSet](https://docs.rs/icu_datetime/0.1.0/icu_datetime/)
+-   Introduce a low-level foundation for string operations with [UnicodeSet](https://docs.rs/icu_codepointset/0.1.0/icu_codepointset/)
```
Comment (sffc) [inline docs/process/prd.md:60]: Suggestion: Change this to go to icu_uniset/0.1.0

File: docs/process/prd.md [hunk 4]
```diff
@@ -117,7 +117,7 @@ Mozilla currently maintains [its own segmentation engine](https://searchfox.org/
 
 Replacing lwbrk with ICU4C for layout needs would require a substantial effort to customize data management to handle Mozilla’s custom data.
 
-ICU4X offers an environment in which we are developing a new Unicode UAX#14/UAX#29 compatible segmentation API which will use the foundational [UnicodeSet API](https://docs.rs/icu_uniset/0.1.0/icu_uniset/), and fit the needs of both Layout and ECMA-402.
+ICU4X offers an environment in which we are developing a new Unicode UAX#14/UAX#29 compatible segmentation API which will use the foundational [CodePointSet API](https://docs.rs/icu_codepointset/0.1.0/icu_codepointset/), and fit the needs of both Layout and ECMA-402.
```
Comment (sffc) [inline docs/process/prd.md:120]: Either revert or link to the staging docs

File: docs/tutorials/writing_a_new_data_struct.md [hunk 5]
```diff
@@ -33,7 +33,7 @@ The data struct definitions should live in the crate that uses them. By conventi
 
 - `icu::decimal::provider::DecimalSymbolsV1`
 - `icu::locale_canonicalizer::provider::LikelySubtagsV1`
-- `icu::uniset::provider::UnicodePropertyV1`
+- `icu::codepointset::provider::UnicodePropertyV1`
```
Comment (sffc) [inline docs/tutorials/writing_a_new_data_struct.md:36]: Nit: This should be icu::properties anyway

File: utils/codepointset/src/lib.rs [hunk 6]
```diff
@@ -2,45 +2,45 @@
 // called LICENSE at the top level of the ICU4X source tree
 // (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
 
-//! `icu_uniset` is a utility crate of the [`ICU4X`] project.
+//! `icu_codepointset` is a utility crate of the [`ICU4X`] project.
 //!
 //! This API provides necessary functionality for highly efficient querying of sets of Unicode characters.
 //!
-//! It is an implementation of the existing [ICU4C UnicodeSet API](https://unicode-org.github.io/icu-docs/apidoc/released/icu4c/classicu_1_1UnicodeSet.html).
+//! It is an implementation of the existing [ICU4C CodePointSet API](https://unicode-org.github.io/icu-docs/apidoc/released/icu4c/classicu_1_1CodePointSet.html).
```
Comment (sffc) [inline utils/codepointset/src/lib.rs:9]: Nit: This one should not change

**Review by sffc — CHANGES_REQUESTED** [review]
Also, please remove the 4th commit from this PR.

File: components/properties/src/api_util.rs [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+impl CodePointSetLike for CodePointSetData {
+    fn contains(&self, ch: char) -> bool {
+        self.payload.get().inv_list.contains(ch)
+    }
+
+    fn contains_u32(&self, code_point: u32) -> bool {
+        self.payload.get().inv_list.contains_u32(code_point)
+    }
+
+    fn iter_ranges(&self) -> Vec<u32> {
```
Comment (sffc) [inline components/properties/src/api_util.rs:39]: This becomes `-> Self::Iter`

File: components/properties/src/api_util.rs [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+    fn contains_u32(&self, code_point: u32) -> bool;
+
+    // return type cannot be `impl ExactSizeIterator<Item = RangeInclusive<u32>>;`. 
+    // compiler says: `impl Trait` not allowed outside of function and method return types
+    // For making FFI easier, we should provide a flat structure of the inversion list 
+    // instead of an iterator of range structs.
+    fn iter_ranges(&self) -> Vec<u32>;
+}
+
+impl CodePointSetLike for CodePointSetData {
```
Comment (sffc) [inline components/properties/src/api_util.rs:30]: Add the fully qualified iterator type here

File: components/properties/src/lib.rs [hunk 9]
```diff
@@ -70,6 +66,9 @@
 
 #![cfg_attr(not(any(test, feature = "std")), no_std)]
 
+extern crate alloc;
```
Comment (sffc) [inline components/properties/src/lib.rs:69]: Don't forget to delete this after removing the Vec

File: components/properties/src/api_util.rs [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+
+//
+// Structs and traits for returning data as sets
+//
+
+pub struct CodePointSetData {
+    pub payload: DataPayload<UnicodePropertyV1Marker>,
+}
+
+pub trait CodePointSetLike {
```
Comment (sffc) [inline components/properties/src/api_util.rs:18]: Issue: This trait should be in `icu_codepointset`

File: .github/workflows/build-test.yml (hunk 2 above)
Comment (sffc) [inline .github/workflows/build-test.yml:471]: Actually, let's NOT change the crate name.  Only change the type name

**Review by sffc — CHANGES_REQUESTED** [review]

File: components/icu/src/lib.rs [hunk 11]
```diff
@@ -371,22 +371,22 @@ pub mod properties {
...
+    //! APIs that return a [`CodePointSet`] exist for binary properties and certain enumerated
     //! properties. See the [`sets`] module for more details.
     //!
     //! APIs that return a [`CodePointTrie`] exist for certain enumerated properties. See the
     //! [`maps`] module for more details.
     //!
     //! # Examples
     //!
-    //! ## Property data as `UnicodeSet`s
+    //! ## Property data as `CodePointSet`s
```
Comment (sffc) [inline components/icu/src/lib.rs:382]: As discussed: CodePointInversionList

**Comment by echeran** [comment]
Closing this old draft PR because it has been superseded by the subsequent proper design work and PRs #2112 and #2140 and other related PRs.

**Pull request closed without merging**